tests/PHPStan/Analyser/data/stringable-object-php8.php

```diff
@@ -0,0 +1,24 @@
+<?php
+
+namespace StringableObjectPhp8;
+
+use function PHPStan\Testing\assertType;
+
+/**
+ * @param stringable-object $object
+ */
+function foo($object)
+{
+	assertType('Stringable', $object);
```

Comment:
This fails (at least locally) because PHPStan is using the Symfony polyfill's definition of `Stringable` instead of the built-in :/

Reply:
Yeah, this isn't great, additionally because of the polyfill PHPStan isn't going to tell us about a nonexistent Stringable pre-PHP 8.0: https://phpstan.org/r/9e44f422-93e6-4923-9911-fe6f75940403

The solution is in this class: https://github.com/phpstan/phpstan-src/blob/1.9.x/src/Reflection/BetterReflection/BetterReflectionSourceLocatorFactory.php

We'd have to debug the ways `Stringable` is discovered before PHP 8 and ideally write a new source locator that would wrap an existing one so that it's not discovered by mistake in a polyfill. It's going to be different in phpstan-src (where the polyfill is known through composer.json) and it's going to be different when PHPStan is installed and used as PHAR (where the polyfill is probably known through AutoloadSourceLocator but also maybe AutoloadFunctionsSourceLocator).

Please note that there's already PhpVersionBlacklistSourceLocator that does this job in some cases.

But at the same time, we can't break projects that use the same polyfill and thus can rely on Stringable already existing before PHP 8.

The previous sentence also means we can't break analysis of phpstan-src itself 🤦‍♂️ 

So what I'd like to see:

1) We need some integration tests around this. Which means to add them here https://github.com/phpstan/phpstan/blob/1.9.x/.github/workflows/other-tests.yml.
  a) Test that runs on PHP 7.4 and references Stringable - expect an error (run with a baseline file and the result should be green - means the error is present)
  b) Test that runs on PHP 8.0 and references Stringable - should be green without a baseline
  c) Test that runs on PHP 7.4, references Stringable, and has polyfills - should be green without a baseline
  d) Test that runs on PHP 8.0, references Stringable, and has polyfills - should be green without a baseline
2) Add a new source locator wrapper for AutoloadSourceLocator so that polyfills from PHPStan's PHAR itself are skipped. Another namespace to skip would be `Hoa` which is unprefixed in the PHAR because it's an old and messy library.
3) Now we have green tests from 1) but we haven't solved our problem in phpstan-src 🤦‍♂️ But that solution doesn't have to be that clean. We can probably come up with some toggle that's used in https://github.com/phpstan/phpstan-src/blob/1.9.x/src/Testing/TestCase.neon and that makes polyfills being skipped in BetterReflectionSourceLocatorFactory even in phpstan-src.